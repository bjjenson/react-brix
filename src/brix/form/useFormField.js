import { useState } from 'react'
import { formatLabel } from './formatLabel'

/**
 * @param  args { import(".").IFormFieldArgs}
 */
export const useFormField = (value, args = {}) => {
  const requiredMessage = args.requiredMessage || 'Required'

  const [state, setState] = useState({
    value,
    error: false,
    helperText: args.helperText,
    pristine: true,
    touched: false,
  })

  const tryValidate = (value, touched) => {
    let result
    if (value && args.validate) {
      result = args.validate(value, args.name)
    }

    if (!Boolean(result) && !args.optional && touched && value === '') {
      result = requiredMessage
    }

    return touched ? {
      error: Boolean(result),
      helperText: result,
    } : {
        error: state.error,
        helperText: state.helperText,
      }
  }

  const handleChange = ({ target }) => {
    const value = args.valueFromTarget ? args.valueFromTarget(target) : target.value
    const coercedValue = args.normalize ? args.normalize(value) : value
    setState({
      ...state,
      value: coercedValue,
      pristine: coercedValue === value,
      ...tryValidate(coercedValue, state.touched),
    })
  }

  const onBlur = () => {
    setState({
      ...state,
      touched: true,
      ...tryValidate(state.value, true),
    })
  }

  const onFocus = () => {
  }

  const setValidationResult = result => {
    setState({
      ...state,
      ...result,
    })
  }

  const validate = () => {
    const result = tryValidate(state.value, true)
    if (result.error) {
      setValidationResult(result)
      return false
    }
    return true
  }

  return {
    props: {
      error: state.error,
      helperText: state.helperText,
      label: formatLabel(args.label, args.optional),
      value: state.value,
      // handlers
      onBlur,
      onChange: handleChange,
      onFocus,
    },
    validate,
    setValidationResult,
  }
}
