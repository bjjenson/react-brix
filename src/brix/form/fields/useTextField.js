import { useFormField } from '../useFormField'

/**
 * @param  args { import("..").IFormFieldArgs}
 * @param  initial String
 */
export const useTextField = (args = {}, initial) => {
  const initialValue = initial || args.value || ''

  return useFormField(initialValue, args)
}
