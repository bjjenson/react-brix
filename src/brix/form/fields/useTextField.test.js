import { useFormField } from '../useFormField'
import { useTextField } from './useTextField'

jest.mock('../useFormField')

let initialArgs
beforeEach(() => {
  initialArgs = {
    optional: false,
    value: 'the value',
    error: 'on',
    helperText: 'the help',
    label: 'the label',
  }

  useFormField.mockReturnValue({ prop1: 'prop1', prop2: 'prop2' })
})

test('sets initial values', () => {
  useTextField(initialArgs)
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('sets initialValue from form', () => {
  useTextField(initialArgs, 'i am initial')
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('defaults value to empty string', () => {
  useTextField()
  expect(useFormField.mock.calls[0]).toMatchSnapshot()
})

test('returns all props needed', () => {
  const actual = useTextField(initialArgs)
  expect(actual).toMatchSnapshot()
})
