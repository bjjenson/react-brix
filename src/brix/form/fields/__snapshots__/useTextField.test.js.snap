// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults value to empty string 1`] = `
Array [
  "",
  Object {},
]
`;

exports[`returns all props needed 1`] = `
Object {
  "prop1": "prop1",
  "prop2": "prop2",
}
`;

exports[`sets initial values 1`] = `
Array [
  "the value",
  Object {
    "error": "on",
    "helperText": "the help",
    "label": "the label",
    "optional": false,
    "value": "the value",
  },
]
`;

exports[`sets initialValue from form 1`] = `
Array [
  "i am initial",
  Object {
    "error": "on",
    "helperText": "the help",
    "label": "the label",
    "optional": false,
    "value": "the value",
  },
]
`;
